### PR TITLE
dbcheck: fix dbcheck crash if password is not set in catalog resource

### PR DESCRIPTION
**Pull request by sebsura**
#### Please check

- [x] Short description and the purpose of this PR is present _above this paragraph_
- [x] Your name is present in the AUTHORS file (optional)

If you have any questions or problems, please give a comment in the PR.

### Helpful documentation and best practices

- [Git Workflow](https://docs.bareos.org/DeveloperGuide/gitworkflow.html)
- [Automatic Sourcecode Formatting](https://docs.bareos.org/DeveloperGuide/generaldevel.html#automatic-sourcecode-formatting)
- [Check your commit messages](https://docs.bareos.org/DeveloperGuide/gitworkflow.html#commits)
- [Boy Scout Rule](https://docs.bareos.org/DeveloperGuide/generaldevel.html#boy-scout-rule)

### Checklist for the _reviewer_ of the PR (will be processed by the Bareos team)
Make sure you check/merge the PR using `devtools/pr-tool` to have some simple automated checks run and a proper changelog record added.

##### General
- [x] Is the PR title usable as CHANGELOG entry?
- [x] Purpose of the PR is understood
- [x] Commit descriptions are understandable and well formatted
- [x] Check backport line
- [x] Required backport PRs have been created

##### Source code quality
- [x] Source code changes are understandable
- [x] Variable and function names are meaningful
- [x] Code comments are correct (logically and spelling)
- [x] Required documentation changes are present and part of the PR

### CODE REVIEW

File: core/src/lib/thread_pool.h [hunk 1]
```diff
@@ -101,9 +101,11 @@ struct thread_pool {
 
   ~thread_pool()
   {
-    for (auto sync : units) { sync->lock()->close(); }
+    for (auto* sync : units) { sync->lock()->close(); }
```
Comment (sebsura) [inline core/src/lib/thread_pool.h:104]: use shared pointers instead ?

File: core/src/win32/compat/compat.cc [hunk 2]
```diff
@@ -470,9 +471,33 @@ static std::wstring AsFullPath(std::wstring_view p)
...
 
-  literal.resize(written);
+    // Sometimes we run into this if the file ended in ' ' or '.' as these
+    // are technically not allowed at the end of paths.  As such
+    // GetFullPathNameW() will not copy them over even tough it included them
+    // in the size computation above (thankfully it does at least that).
+    // As we do want to support file names like that we have to make sure
+    // that we did not run into a situation like that and, if we did, we need
+    // to fixup the situation.
+    if (written < required) {
```
Comment (sebsura) [inline core/src/win32/compat/compat.cc:482]: test with other weird characters in character names

**Comment by sebsura** [comment]
Need to fix the bahavior with invalid directory names as well!

File: core/src/lib/thread_pool.h [hunk 3]
```diff
@@ -173,29 +173,33 @@ struct thread_pool {
...
 
-  std::vector<synchronized<work_unit>*> units;
+  std::vector<std::unique_ptr<synchronized<work_unit>>> units;
   std::vector<std::thread> threads{};
 
   void add_thread()
   {
-    auto* sync = new synchronized<work_unit>();
-    units.push_back(sync);
+    // sync will live at least as much as threads, so its ok to not use
```
Comment (pstorz) [inline core/src/lib/thread_pool.h:181]:
```suggestion
    // sync will live at least as long as threads, so its ok to -*NOT* use
```

**Review by pstorz — APPROVED** [review]
Please check comments

**Pull request merged**